Fix emit path hashing

jsgf opened the pull request:
With `--emit KIND=PATH`, the PATH should not affect hashes used for dependency tracking. It does not with other ways of specifying output paths (`-o` or `--out-dir`).

Also updates `rustc -Zls` to print more info about crates, which is used here to implement a `run-make` test.

It seems there was already a test explicitly checking that `OutputTypes` hash *is* affected by the path. I think this behaviour is wrong, so I updated the test.

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

bjorn3 commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -601,10 +601,23 @@ impl MetadataBlob {
+                dep.hash,
+                dep.host_hash,
+                dep.kind
+            )?;
```
Nice!

jackh726 commented on compiler/rustc_session/src/config.rs:
```diff
@@ -325,11 +329,26 @@ impl Default for TrimmedDefPaths {
+#[derive(Clone, Debug)]
 pub struct OutputTypes(BTreeMap<OutputType, Option<PathBuf>>);
 
-impl_stable_hash_via_hash!(OutputTypes);
```
Does this just forward the HashStable impl to Hash? Theoretically it can stay then, right? Since Hash is still implemented, just not derived?

jsgf replied:
Ah, yeah. I changed the stable hash first thinking it was what would be used for the dep hash, but I never saw it being called so I changed the Hash impl, but didn't think to revert this change. I'll flip it back.

jackh726 reviewed (approved):
Seems simple enough to me

jsgf commented:
Switched back to using `impl_stable_hash_via_hash!(OutputTypes);`

Aaron1011 commented:
@jsgf: Can you elaborate on the motivation for this change? While it doesn't *appear* that any queries are depending on the path in `OutputTypes`, nothing prevents someone from adding such a dependency in the future.

Absent a strong reason (e.g. the option is changed frequently, and therefore impacts incremental compilation time), I think we should not be removing information from the dep-tracking hash.

jsgf commented:
@Aaron1011 There's no reason to include the path in the hash. If you were to consider it important, then the `-o` and `--out-dir` options should also be tracked. You would also have to have a rationale for why `rustc --emit link=path1 ...` and `rustc --emit link=path2 ...; mv path2 path1` should generate different artifacts.

The specific motivation is that I'm trying to use `--emit link=path1` and `--emit link=path2 -Zno-codegen` to approximate `--emit link,metadata` - that is, pipelined builds but using two invocations of rustc. If the two different `--emit link=path` operations generate different hashes, then this doesn't work.

Fundamentally, dep tracking is trying to track the content of the emitted artifacts, and specifically, distinguish incompatible artifacts to prevent them from being used together. The actual pathname used to write the artifact to doesn't have any role in this.

Aaron1011 commented:
@jsgf: In addition to being used as input for the crate hash, the `DepTrackingHash` is also used when determining whether or not to re-use the previous incremental compilation state.

I agree that the crate hash shouldn't depend on the path passed to `--emit` - however, the incremental hash should still depend on it. I think we'll need to modify the `options!` macro and/or the `DepTrackingHash` trait - we'll want to hash just the keys when computing the crate hash, but both the keys and values when computing the incremental hash (this is controlled by the `for_crate_hash` parameter).

I'm sorry about the extra complexity here. However, we've recently uncovered a large number of ICEs due to improper tracking of global state during incremental compilation, and I'd like to avoid regressing our current handling.

Feel free to message me on Zulip if you'd like to discuss the changes further.

bjorn3 commented:
> I agree that the crate hash shouldn't depend on the path passed to --emit - however, the incremental hash should still depend on it.

No, it only depends on which artifacts are requested, not where they are written. Even with incr comp the artifact destination is not assumed to already have the artifact when nothing has changed, so the artifacts are copied every time to the current artifact destination.

Aaron1011 commented:
@bjorn3: Nothing actually enforces that requirement - we've previously ran into issues where queries ended up depending on the value/order of something that "shouldn't" have mattered , but did (`-l`) flags.

I strongly believe that in order to avoid re-introducingrhe same type of bugs that caused us to disable incremental compilation recently, we need to remove/restrict untracked global state. If we aren't going to hash the filename, then we should somehow ensure that it can't be accessed via `tcx`.

jsgf commented:
@Aaron1011 It's not clear to me whether you're pointing out a specific problem with this PR, or raising an issue of a general class of problems that overlap with this PR's concerns.

From my POV, the intent is to make all the ways of naming output files functionally equivalent; it seems absurd to me that you'd get different bitwise output for `--emit link -o libfoo.rlib` and `--emit link=libfoo.rlib`, and that the resulting .rlib files are not interchangable. Do you have an objection to fixing this discrepency now, and then addressing your broader issues separately?

jsgf commented:
Alternatively, we could make OutputTypes TRACKED_BUT_NO_CRATE_HASH, which I think addresses both my and @Aaron1011's concerns. It would also have the effect of making OutputType not affect the crate hashes at all, which I think is even better. After all, why should `--emit link,dep-info` and two invocations of `--emit link` and `--emit dep-info` generate different crate hashes? The dep-info is immaterial to the rlib. The only awkward thing here is that `--emit link,metadata` generates different .rmeta files from `--emit metadata`. But from a dependency tracking POV I think that's fine - it's strictly a superset, and the content that's common is (should be?) semantically identical.

bjorn3 commented:
> it's strictly a superset, and the content that's common is (should be?) semantically identical.

`dependency_format::calculate_type` returns an empty array for `--emit metadata`. AFAICT the result of this method depends on the result of this function encoded in the metadata of dylibs. I wouldn't be surprised if this leads to problems when mixing `--emit metadata` metadata with `--emit link` dylibs.

jsgf commented:
@bjorn3 Hm, I'll look into that. But in general I've found `--emit link -Zno-codegen` a much more reliable way to emit metadata than `--emit metadata`.

michaelwoerister commented:
Making this `TRACKED_BUT_NO_CRATE_HASH` might be a good idea -- but please be careful with changes here. Let me know if I should take a closer look.

jackh726 commented:
Let's r? @Aaron1011 since they're in a better position to review this than I am

jsgf commented:
@Aaron1011 @michaelwoerister I updated this to use `TRACK_NO_CRATE_HASH` for `OutputTypes`. I left the tests I added unchanged, and there's no regressions in those or any other tests I could see.

@bjorn3 I also made `-Zno-codegen` `TRACK_NO_CRATE_HASH` since it shouldn't affect the crate content with respect to dependency tracking.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2788883593) [(plain)](https://github.com/rust-lang/rust/commit/2728ba0170fecbd7864d1529dd8394837b2dcdf9/checks/2788883593/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  left: `1029918945431836347`,
 right: `8612613394021905856`', compiler/rustc_interface/src/tests.rs:85:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::test_debugging_options_tracking_hash stdout ----
thread 'tests::test_debugging_options_tracking_hash' panicked at 'assertion failed: `(left != right)`
 right: `10200085699698919768`', compiler/rustc_interface/src/tests.rs:92:5


failures:
failures:
    tests::test_debugging_options_tracking_hash
    tests::test_output_types_tracking_hash_different_paths

test result: FAILED. 13 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s

error: test failed, to rerun pass '-p rustc_interface --lib'


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" " llvm" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc_interface" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:22:41
```

</details>

bjorn3 commented on compiler/rustc_session/src/options.rs:
```diff
@@ -132,7 +132,7 @@ top_level_options!(
         force_warns: Vec<String> [TRACKED],
         describe_lints: bool [UNTRACKED],
-        output_types: OutputTypes [TRACKED],
+        output_types: OutputTypes [TRACKED_NO_CRATE_HASH],
```
I am concerned about the fact that changing which output types are requested doesn't change the crate hash due to `if output_types.should_codegen()` occurring in several locations that can affect the crate metadata like the list of reachable non-generics, the list of exported symbols and `dependency_format::calculate_type`. In principle these `should_codegen()` fast-paths could be removed, but it will likely regress `cargo check` performance.

jsgf replied:
In https://internals.rust-lang.org/t/rfc-emit-check/14711/5?u=jsgf I measured it as ~8% difference on `cargo check` for Cargo itself. I don't know how that relates to anything else.

I experimented with making a distinction between `--emit metadata` and `--emit check` and corresponding `.rmeta` and `.rcheck` files, but the addition of a new file type in the mix made things complex in location resolution in rustc and in Cargo itself. It would be much easier to just make `.rmeta` invariant.

One advantage, in principle, is that if `cargo check`'s emitted metadata is "full fat", then it can be used immediately from cache in `cargo build` which would effectively make the dependency graph flat and allow the rlibs to be generated completely in parallel - though I expect that would only really be interesting for builds which are closer to build from scratch vs incremental rebuild.

jsgf replied:
I also wonder what the practical effect of this will be. I presume the risk is that some build system - like `cargo` will intermix invocations of `--emit metadata` (cargo check) and `--emit link,metadata` (cargo build) and get failures if check metadata gets used for builds.

But what's the actual outcome? In the worst case, it could appear to succeed but generate subtly broken output. But I don't think that's the case here, is it? It will fail to find " list of reachable non-generics, the list of exported symbols and dependency_format::calculate_type", and fail as a result. Vs if the hashes are different, it will fail for that reason ("found possibly newer version of crate ...").

(If need be, could we extend the .rmeta format to include a type to explicitly flag "for check use only"?)

I haven't looked into how `cargo` specifically managed check vs build .rmetas, but I think it avoids the problem by always regenerating .rmetas as part of a build. In my own experiments with Buck, I make sure to put them in separate dirs (which is the motivation for this PR, since currently doing that changes the crate hash independent of anything else).

Either way, any failure is a bug in the build system, not something that an end-user would expect to see with a correctly implemented build system.

bjorn3 replied:
> I also wonder what the practical effect of this will be. I presume the risk is that some build system - like cargo will intermix invocations of --emit metadata (cargo check) and --emit link,metadata (cargo build) and get failures if check metadata gets used for builds.

That is exactly what you want to do, right? Using the `--emit metadata` output of dependencies to compile `--emit link` crates.

> But what's the actual outcome? In the worst case, it could appear to succeed but generate subtly broken output. But I don't think that's the case here, is it? It will fail to find " list of reachable non-generics, the list of exported symbols and dependency_format::calculate_type", and fail as a result. Vs if the hashes are different, it will fail for that reason ("found possibly newer version of crate ...").

A wrong result of dependency_format::calculate_type will cause linking to silently fail due to mixing up static and dynamic linking of crates. For example trying to statically link a crate that is already linked into a dynamic library dependency.

A wrong result for the list of reachable non-generics and exported symbols would likely result in the mono item collector creating mono items that are already included in dependencies (or maybe omitting necessary ones), causing duplicate linker errors.

> I haven't looked into how cargo specifically managed check vs build .rmetas, but I think it avoids the problem by always regenerating .rmetas as part of a build.

It currently handles check and regular builds independently, but we want `cargo check` to reuse `cargo build` rmeta files in the future.

bjorn3 replied:
Yes, just the types being hashes is fine. `-Zno-codegen` doesn't affect the metadata in anyway AFAIK, so not hashing it for the crate hash is fine.

jsgf replied:
> That is exactly what you want to do, right? Using the --emit metadata output of dependencies to compile --emit link crates.

If `--emit link` and `--emit metadata` inherently generate different crate hashes, then that's useless to me. I currently have to use `--emit link -Zno-codegen` to generate complete hash-compatible metadata. That's fine for experiments, but I'd like to get to something that can be stabilized, and ideally directly expresses the intent.

My current opinion is that using `--emit metadata` which always generates the same metadata regardless of other emits which is crate-hash compatible with `--emit link` would be ideal, but I'm open to any suggestion that meets the goals of:
- separate rustc invocations to generate compatible metadata and codegen
- on a stabliization track

bjorn3 replied:
> My current opinion is that using --emit metadata which always generates the same metadata regardless of other emits which is crate-hash compatible with --emit link would be ideal

I agree, but probably not everyone may agree with the associated perf loss unfortunately.

oxalica replied:
> the associated perf loss

Just a question, is it possible to reuse these fat rmeta generated in checking (seems it contains MIR and type information) when building binaries? If so, we can save 100% of check time (the time populating rmeta) during the build.
But sadly we currently just re-work on analysis and re-generate rmeta again when `--emit link`.

bjorn3 replied:
Yes, build -> check reusing is already possible, but simply not implemented in cargo. There should be an issue open for that. check -> build reusing won't work even with this PR.

oxalica replied:
Is check -> build reusing technically possible? I'm not sure if everything needed for codegen is already available in rmeta.

bjorn3 replied:
In theory it would be possible to never exclude anything from the rmeta even when no codegen is required. This would allow using the check rmeta for builds, but still require all crates to be compiled with codegen. The benefit of this would be more parallelization as codegen for all crates can immediately start without having to wait for the compilation of dependencies to get metadata. The downside is that the check metadata becomes larger which slows down check runs.

jsgf commented:
What's the next step for this? It's approved but still tagged "waiting on review". Does @Aaron1011 also need to approve?

bjorn3 commented:
Yes, he is assigned to this PR and I would like someone else to look at it anyway just in case.

jsgf commented:
Rebased, and added a commit from @Aaron1011 which makes `OutputTypes` a TRACKED/TRACKED_NO_CRATE_HASH hybrid (keys are tracked, paths are no crate hash). @bjorn3 could you review this; in conversation @Aaron1011 said he's fine with this PR with this addition but wanted another reviewer for it.

bjorn3 commented on src/test/run-make/emit-path-unhashed/Makefile:
```diff
@@ -0,0 +1,26 @@
+
+OUT=$(TMPDIR)/emit
+
+# --emit KIND=PATH should not affect crate hash vs --emit KIND
```
Can you check that adding or removing `--emit` kinds does change the crate hash?

jsgf replied:
OK, done.

bjorn3 reviewed (commented):
r=me with the extra test

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 45146978e8d593d5ec713f54140407d64f620800 has been approved by `bjorn3`

<!-- @bors r=bjorn3 45146978e8d593d5ec713f54140407d64f620800 -->
<!-- homu: {"type":"Approved","sha":"45146978e8d593d5ec713f54140407d64f620800","approver":"bjorn3"} -->

bors commented:
:hourglass: Testing commit 45146978e8d593d5ec713f54140407d64f620800 with merge b8be3162d734f3583b240977615f3e1bae6b364a...
<!-- homu: {"type":"BuildStarted","head_sha":"45146978e8d593d5ec713f54140407d64f620800","merge_sha":"b8be3162d734f3583b240977615f3e1bae6b364a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2888778239)
Approved by: bjorn3
Pushing b8be3162d734f3583b240977615f3e1bae6b364a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2888778239"},"merge_sha":"b8be3162d734f3583b240977615f3e1bae6b364a"} -->
